javascript/src/URDFClasses.js

```diff
@@ -242,9 +246,81 @@ class URDFJoint extends URDFBase {
+                    // Respect origin RPY when setting position
+                    if (posX !== this.jointValue[0] && posX !== null) {
+                        _tempAxis.set(1, 0, 0).applyEuler(this.rotation);
+                        this.position.addScaledVector(_tempAxis, posX);
+                        didUpdate = true;
+                    }
+                    if (posY !== this.jointValue[1] && posY !== null) {
+                        _tempAxis.set(0, 1, 0).applyEuler(this.rotation);
+                        this.position.addScaledVector(_tempAxis, posY);
+                        didUpdate = true;
+                    }
+                    if (posZ !== this.jointValue[2] && posZ !== null) {
+                        _tempAxis.set(0, 0, 1).applyEuler(this.rotation);
+                        this.position.addScaledVector(_tempAxis, posZ);
+                        didUpdate = true;
+                    }
+
+                    // Doing the math to set each individual Euler value seemed more awkward than just handling all this in the same case using ternaries.
+                    if (roll !== this.jointValue[3] || pitch !== this.jointValue[4] || yaw !== this.jointValue[5]) {
+                        this.jointValue[3] = roll !== null ? roll : this.jointValue[3];
+                        this.jointValue[4] = pitch !== null ? pitch : this.jointValue[4];
+                        this.jointValue[5] = yaw !== null ? yaw : this.jointValue[5];
+                        this.quaternion.setFromEuler(
+                            _tempEuler.set(
+                                this.jointValue[3],
+                                this.jointValue[4],
+                                this.jointValue[5],
+                                'XYZ',
+                            ),
+                        ).premultiply(this.origQuaternion);
+                        didUpdate = true;
+                    }
+                }
+
+                this.matrixWorldNeedsUpdate = didUpdate;
+                return didUpdate;
+
             case 'planar':
-                // TODO: Support these joint types
-                console.warn(`'${ this.jointType }' joint not yet supported`);
+                // no-op if all values are identical to existing value or are null
+                if (this.jointValue.every((value, index) => values[index] === value || values[index] === null)) return didUpdate;
+
+                // Planar joints have three degrees of freedom: X distance, Y distance, Z rotation.
+                const [posX, posY, rotZ] = values;
+
+                // Respect existing RPY when modifying the position of the X,Y axes
+                this.position.copy(this.origPosition);
+                if (posX !== null) {
+                    _tempAxis.set(1, 0, 0).applyEuler(this.rotation);
+                    this.position.addScaledVector(_tempAxis, posX);
+                    this.jointValue[0] = posX;
```

Comment:
From the specification the plane that the joint is supposed to move on should be based on the axis specified in the joint - this is just using the local x axis:

> planar — this joint allows motion in a plane perpendicular to the axis.

One additional issue with the spec is that it does not clearly define how to derive the X and Y plane axes from the single "axis" vector. It could just be from rotating the frame such that the z axis the shortest direction to the axis direction with no twist rotation - but this leaves us with an undefined case when the axis is `0 0 - 1`. If there is a consistent and well-defined approach to this in MoveIt or other applications we should use that one. It would also be good to update the specification document to be more clear here.